src/main/java/org/testng/TestRunner.java

```diff
@@ -1064,7 +1065,10 @@ private void afterRun() {
     // end up creating cycles when combined with preserve-order.
     boolean hasDependencies = false;
     for (ITestNGMethod m : methods) {
-      result.addNode(m);
+      boolean added = result.addNode(m);
+      if (!added) {
+        result.addNode(new WrappedTestNGMethod(m));
```

Comment:
What if we always use WrappedTestNGMethod?

Reply:
> What if we always use WrappedTestNGMethod?

Not sure if its needed. The only time wherein I thought we needed this was when our users used a method interceptor, which internally bumped up the number of test methods (Please refer [here](https://github.com/cbeust/testng/blob/master/src/test/java/test/verify/VerifyTest.java#L22-L24) )

Reply:
I think it will help the readability and simplify the code if we have only one generic workflow.
I think some comments will help too.

Reply:
I will add comments and push it in once I go home. But we still need the conditional insertion of `WrappedTestNGMethod` because if we dont do it, then we end up with the original situation wherein methods are being lost. Here's a scenario 

Lets say I have 2 methods and a user used a Method interceptor and then changed it to 3 methods by duplicating one of the methods. Now if we wrap all methods within the `WrappedTestNGMethod` then our logic will fail when we are attempting at adding the new method instance which was literally cloned because it would have the same `hashCode` (our implementation just falls back to the hashcode value of `Method` which in this case would be the same). That is why we first need to try injecting the `TestNGMethod` instance. if its injection fails then we try and inject it via `WrappedTestNGMethod` (which internally fudges the hashCode and ensures that it would be different). I am thinking that I should perhaps change that constant `31` and instead use a random number as the multiplication factor, because I am thinking that it will still cause problems if the user attempts duplicating the same method multiple times and then we try creating a `DynamicGraph` out of it.

Reply:
Great! I think the test cases you described should be added too ;)

Reply:
@juherr - Sure thing. I think that test case itself came from [this](http://beust.com/weblog/2010/03/23/better-mock-testing-with-testng/) blog post from @cbeust (I remember that `@Verify` annotation from his blog only). This time I will add these changes as an additional commit so that its easy for vetting out.